python/vowpalwabbit/pyvw.py

```diff
@@ -1015,24 +1015,30 @@ def push_features(self, ns, featureList):
         self.ex.push_features(self.ns, featureList)
 
 
-class abstract_label:
+class AbstractLabel:
     """An abstract class for a VW label."""
 
     def __init__(self):
         pass
 
-    def from_example(self, ex):
+    def from_example(self, ex: pylibvw.example):
```

Comment:
Dont we want this to be the Python type? `Example`?

Reply:
The issue here is that `example` is defined after `SimpleLabel`, and `example` uses `SimpleLabel` as a default value so it has to be defined after

Reply:
This forward declaration problem has been solved in Python 3.7 (https://www.python.org/dev/peps/pep-0563/#enabling-the-future-behavior-in-python-3-7) using `from __future__ import annotations` to delay evaluation of annotations. We are still supporting Python3.6 though. Python 3.6 reached end of support as of December 2021, so perhaps we can use this soon.

I think until we can use that another option is to write the annotation as a string literal (see here: https://www.python.org/dev/peps/pep-0484/#id28). Pytype still evaluates this correctly as far as I can tell and it allows for use of a later defined type

```suggestion
    def from_example(self, ex: "example"):
```

Reply:
Cool added in the string literal